File: Python/113 路径总和II.py
"""
113. 路径总和 II
给定一个二叉树和一个目标和，找到所有从根节点到叶子节点路径总和等于给定目标和的路径。

说明: 叶子节点是指没有子节点的节点。

示例:
给定如下二叉树，以及目标和 sum = 22，

              5
             / \
            4   8
           /   / \
          11  13  4
         /  \    / \
        7    2  5   1
返回:

[
   [5,4,11,2],
   [5,8,4,5]
]
通过次数40,021提交次数67,765

"""
import tools
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class mySolution:
    def pathSum(self, root: TreeNode, sum: int):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return []
        res = []
        def LDR (root,nowSum,path):
            if root == None:
                return 
            nowSum += root.val
            path.append(root.val)
            if root.left is None and root.right is None:
                if nowSum == sum:
                    res.append(path)
            else:
                LDR(root.left,nowSum,path[:])
                LDR(root.right,nowSum,path[:])
        
        LDR(root,0,[])
        return res

#大佬思路 路径回溯 不用传递路径参数
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return []
        if root.left==None and root.right==None and root.val==sum:
            return [[root.val]]
        result=[]
        lefttree=self.pathSum(root.left,sum-root.val)
        righttree=self.pathSum(root.right,sum-root.val)
        for i in lefttree+righttree:
            result.append([root.val]+i)
        return result
s = Solution()
tr = tools.stringToTreeNode("[5,4,8,11,null,13,4,7,2,null,null,5,1]")
print(s.pathSum(tr,22))

File: CPP/113 路径总和II.cpp
#include "include.hpp"
using namespace std;
/*
113. 路径总和 II
给定一个二叉树和一个目标和，找到所有从根节点到叶子节点路径总和等于给定目标和的路径。

说明: 叶子节点是指没有子节点的节点。

示例:
给定如下二叉树，以及目标和 sum = 22，

              5
             / \
            4   8
           /   / \
          11  13  4
         /  \    / \
        7    2  5   1
返回:

[
   [5,4,11,2],
   [5,8,4,5]
]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    public:
    vector<vector<int>> res{};
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        pathSumPart(root,sum,vector<int>());
        return res;
    }

    void pathSumPart(TreeNode* root, int sum,vector<int>& p) {
        if(root == nullptr) return;
        p.push_back(root->val);
        if(root->left == nullptr && root->right== nullptr && sum-root->val == 0)
        {
            res.push_back(p);
        }
        pathSumPart(root->left,sum - root->val,p);
        pathSumPart(root->right,sum - root->val,p);
        p.pop_back();
    }
};

class Solution {
    public:
    vector<vector<int>> res{};
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        return  *pathSumPart(root,sum);
    }

    vector<vector<int>>* pathSumPart(TreeNode* root, int sum) {
        if(root == nullptr) return nullptr;
        if(root->left == nullptr && root->right== nullptr && sum-root->val == 0)
        {
            return new vector<vector<int>>{vector<int>{root->val}};
        }
        vector<vector<int>>* pl =  pathSumPart(root->left,sum - root->val);
        vector<vector<int>>* pr = pathSumPart(root->right,sum - root->val);
        vector<vector<int>>* rtn = new vector<vector<int>>();
        if(!pl)
        {
            for(auto x:*pl)
            {
                x.push_back(root->val);
                rtn->push_back(x);
            }
        }

        if(!pr)
        {
            for(auto x:*pr)
            {
                x.push_back(root->val);
                rtn->push_back(x);
            }
        }

        return rtn;
    }
};

int main()
{
    Solution s;
    system("pause");
};